jerry-core/ecma/builtin-objects/ecma-builtin-helpers-sort.c

```diff
@@ -0,0 +1,146 @@
+                                   const sort_compare_t sort_cb) /**< sorting callback */
+{
+  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
+
+  /* Left child of the current index. */
+  uint32_t child = index * 2 + 1;
+  ecma_value_t swap = array_p[index];
+  bool should_break = false;
+
+  while (child <= right && ecma_is_value_empty (ret_value) && !should_break)
+  {
+    if (child < right)
+    {
+      /* Compare the two child nodes. */
+      ECMA_TRY_CATCH (child_compare_value, sort_cb (array_p[child], array_p[child + 1], compare_func),
+                      ret_value);
+
+      JERRY_ASSERT (ecma_is_value_number (child_compare_value));
+
+      /* Use the child that is greater. */
+      if (ecma_get_number_from_value (child_compare_value) < ECMA_NUMBER_ZERO)
+      {
+        child++;
+      }
+
+      ECMA_FINALIZE (child_compare_value);
+    }
+
+    if (ecma_is_value_empty (ret_value))
+    {
+      JERRY_ASSERT (child <= right);
+
+      /* Compare current child node with the swap (tree top). */
+      ECMA_TRY_CATCH (swap_compare_value, sort_cb (array_p[child], swap, compare_func), ret_value);
+      JERRY_ASSERT (ecma_is_value_number (swap_compare_value));
+
+      if (ecma_get_number_from_value (swap_compare_value) <= ECMA_NUMBER_ZERO)
+      {
+        /* Break from loop if current child is less than swap (tree top) */
+        should_break = true;
+      }
+      else
+      {
+        /* We have to move 'swap' lower in the tree, so shift current child up in the hierarchy. */
+        uint32_t parent = (child - 1) / 2;
+        JERRY_ASSERT (parent <= right);
+        array_p[parent] = array_p[child];
+
+        /* Update child to be the left child of the current node. */
+        child = child * 2 + 1;
+      }
+
+      ECMA_FINALIZE (swap_compare_value);
+    }
+  }
+
+  /*
+   * Loop ended, either current child does not exist, or is less than swap.
+   * This means that 'swap' should be placed in the parent node.
+   */
+  uint32_t parent = (child - 1) / 2;
+  JERRY_ASSERT (parent <= right);
+  array_p[parent] = swap;
+
+  if (ecma_is_value_empty (ret_value))
+  {
+    ret_value = ECMA_VALUE_UNDEFINED;
+  }
+
+  return ret_value;
+} /* ecma_builtin_helper_array_to_heap */
+
+/**
+ * Heapsort function
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+ecma_value_t
+ecma_builtin_helper_array_heap_sort_helper (ecma_value_t *array_p, /**< array to sort */
+                                            uint32_t right, /**< right index */
+                                            ecma_value_t compare_func, /**< compare function */
+                                            const sort_compare_t sort_cb) /**< sorting callback */
+{
+  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
+
+  /* First, construct the ordered binary tree from the array. */
+  for (uint32_t i = (right / 2) + 1; i > 0 && ecma_is_value_empty (ret_value); i--)
```

Comment:
To allow having the type of the index a `uint32_t` I made a small change here where I added one more to `i` and changed the condition from `i >= 0` to `i > 0` because before you would decrement when `i = 0` (which would previously terminate the loop) but after making the value unsigned this would wrap to the max unsigned value and so the loop would never terminate.